### PR TITLE
Remove some unnecessary check logic for lang items in HIR typeck

### PR DESCRIPTION
**Pull request by compiler-errors**
Obvious bugs with `#[no_core]` do not deserve customized recovery logic, since they are bugs we do not expect users to ever encounter, and if users are experimenting with `#[no_core]`, they should really be familiar with the compiler implementation.

These error recoveries are implemented now only where issues have been reported in the past, rather than systematically validating lang items.

See https://github.com/rust-lang/compiler-team/issues/620
> In particular, one-off fixes for particular assumptions about lang items or intrinsics that introduce additional complexity into the compiler are not accepted.

r? Nilstrieb

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 1]
```diff
@@ -255,11 +255,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             ) {
                 // Check for `self` receiver on the method, otherwise we can't use this as a `Fn*` trait.
                 if !self.tcx.associated_item(ok.value.def_id).fn_has_self_parameter {
```
Comment (Nilstrieb) [inline compiler/rustc_hir_typeck/src/callee.rs:257]: Can the entire check be removed?

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 2]
```diff
@@ -948,9 +938,7 @@ impl<'a, 'tcx> DeferredCallResolution<'tcx> {
                 );
             }
             None => {
-                // This can happen if `#![no_core]` is used and the `fn/fn_mut/fn_once`
-                // lang items are not defined (issue #86238).
-                fcx.dcx().emit_err(errors::MissingFnLangItems { span: self.call_expr.span });
+                bug!("")
```
Comment (Nilstrieb) [inline compiler/rustc_hir_typeck/src/callee.rs:941]:
```suggestion
                bug!("missing n/fn_mut/fn_once lang items")
```
Adding a message also makes it easier to understand why this happens when reading the code.

File: compiler/rustc_hir_typeck/src/check.rs [hunk 3]
```diff
@@ -152,26 +152,22 @@ pub(super) fn check_fn<'a, 'tcx>(
...
     if generic_counts.consts != 0 {
-        let span = tcx.def_span(fn_id);
         tcx.dcx().span_err(span, "should have no const parameters");
     }
 
-    let Some(panic_info_did) = tcx.lang_items().panic_info() else {
-        tcx.dcx().err("language item required, but not found: `panic_info`");
-        return;
-    };
+    let panic_info_did = tcx.require_lang_item(hir::LangItem::PanicInfo, Some(span));
```
Comment (Nilstrieb) [inline compiler/rustc_hir_typeck/src/check.rs:170]: A prettier error message by making the code shorter, beautiful ferrisowo

File: compiler/rustc_hir_typeck/src/callee.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/callee.rs:941]: Oops, I thought I had put TODO here to fill in later, but I think I just left it empty.

File: compiler/rustc_hir_typeck/src/callee.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/callee.rs:257]: Yeah true

**Comment by Nilstrieb** [comment]
r=me when either CI or you are happy

**Pull request merged**